### PR TITLE
RR-738 - Initial switch to short question set

### PR DESCRIPTION
**Pull request by mattwills-moj-digital**
This PR starts the journey when switching from the "long" route to the "short" one. In other words, when the prisoner was previously aiming to work after release, but is no longer hoping to. This PR gets us as far as adding an individual qualification or additional training.

One of the main focuses in this PR has been to dynamically resolve the back links, so that they account for when updating an individual aspect of an induction, or when they are part of a journey to switch between the long/short question sets. To achieve this, I've introduced `pageFlowHistory`, which is built up and positioned as the user works their way through a flow.

### CODE REVIEW

File: server/@types/express/index.d.ts [hunk 1]
```diff
@@ -34,10 +40,12 @@ declare module 'express-session' {
     newGoals: Array<NewGoal> // An array of NewGoal representing the Goals that have been added
     updateGoalForm: UpdateGoalForm
     prisonerListSortOptions: string
-    pageFlowQueue: PageFlowQueue
+    pageFlowHistory: PageFlow
```
Comment (mattwills-moj-digital) [inline server/@types/express/index.d.ts:43]: This allows us to keep a separate list of pages that have been visited (rather than trying to repurpose the `pageFlowQueue` which got messy, especially when it came to the previous work experience!).

File: server/@types/viewModels/index.d.ts [hunk 2]
```diff
@@ -414,8 +414,12 @@ declare module 'viewModels' {
     prisonName: string
   }
 
-  export interface PageFlowQueue {
+  export interface PageFlow {
```
Comment (mattwills-moj-digital) [inline server/@types/viewModels/index.d.ts:417]: Renamed, since this is used by both the `PageFlowQueue` and `PageFlowHistory` modules.

File: server/routes/induction/update/previousWorkExperienceTypesUpdateController.ts [hunk 3]
```diff
@@ -11,18 +11,24 @@ import TypeOfWorkExperienceValue from '../../../enums/typeOfWorkExperienceValue'
...
+    const { pageFlowHistory } = req.session
+    if (pageFlowHistory) {
+      return getPreviousPage(pageFlowHistory)
+    }
     const { prisonNumber } = req.params
     return `/plan/${prisonNumber}/view/work-and-interests`
   }
 
   getBackLinkAriaText(req: Request): string {
+    // TODO - retrieve text for previous work experience types (a generic one for all types?)
```
Comment (mattwills-moj-digital) [inline server/routes/induction/update/previousWorkExperienceTypesUpdateController.ts:31]: There's a little bit of complexity to work on here...

File: server/routes/induction/update/qualificationDetailsUpdateController.ts [hunk 4]
```diff
@@ -47,10 +47,14 @@ export default class QualificationDetailsUpdateController extends QualificationD
       qualificationLevelForm.qualificationLevel,
     )
 
+    if (req.session.updateInductionQuestionSet) {
+      // TODO redirect to next page in the question set
+      return res.redirect(`/prisoners/${prisonNumber}/induction/qualifications`)
```
Comment (mattwills-moj-digital) [inline server/routes/induction/update/qualificationDetailsUpdateController.ts:52]: This is as far as I've got. We need to determine the next page here.

File: server/routes/induction/update/qualificationsListUpdateController.ts [hunk 5]
```diff
@@ -15,6 +15,18 @@ export default class QualificationsListUpdateController extends QualificationsLi
 
   getBackLinkUrl(req: Request): string {
     const { prisonNumber } = req.params
+    const { pageFlowHistory } = req.session
+    if (pageFlowHistory) {
+      // TODO - this needs thinking through!
```
Comment (mattwills-moj-digital) [inline server/routes/induction/update/qualificationsListUpdateController.ts:20]: I got to this late last night and couldn't test it due to the server instances being shutdown! 😂 However, that may be a blessing in disguise as this feels somewhat messy/brittle. I'll have a fresh look at this tomorrow 👍

File: server/routes/induction/update/additionalTrainingUpdateController.ts [hunk 6]
```diff
@@ -51,6 +56,11 @@ export default class AdditionalTrainingUpdateController extends AdditionalTraini
     }
 
     const updatedInduction = this.updatedInductionDtoWithAdditionalTraining(inductionDto, additionalTrainingForm)
+
+    if (req.session.updateInductionQuestionSet) {
+      // TODO redirect to next page in the question set
```
Comment (mattwills-moj-digital) [inline server/routes/induction/update/additionalTrainingUpdateController.ts:61]: This is as far as I've got (along with the branch to add a qualification). We need to determine the next page here.

File: server/routes/induction/common/inductionController.ts [hunk 7]
```diff
@@ -13,4 +14,12 @@ export default abstract class InductionController {
    * Concrete classes must provide an implementation of this method to provide the backlink aria text relevant to their use case / journey.
    */
   abstract getBackLinkAriaText(req: Request): string
+
+  addCurrentPageToHistory(req: Request, currentPageUri: string) {
```
Comment (nathanrussell-moj-digital) [inline server/routes/induction/common/inductionController.ts:18]: Good idea to put this in the superclass 👍 
Nice

File: server/routes/induction/common/previousWorkExperienceDetailController.ts [hunk 8]
```diff
@@ -34,6 +34,14 @@ export default abstract class PreviousWorkExperienceDetailController extends Ind
       return next(createError(404, `Previous Work Experience type ${typeOfWorkExperience} not found on Induction`))
     }
 
+    // Check if we are in the midst of changing the main induction question set (e.g. from long route to short route)
```
Comment (nathanrussell-moj-digital) [inline server/routes/induction/common/previousWorkExperienceDetailController.ts:37]: Not that I'm asking you to change this comment, but would it still be true if we were changing from short to long? I'm assuming we'd still need to do this?

File: server/routes/induction/common/wantToAddQualificationsController.ts [hunk 9]
```diff
@@ -19,6 +19,9 @@ export default abstract class WantToAddQualificationsController extends Inductio
     next: NextFunction,
   ): Promise<void> => {
     const { prisonerSummary, prisonerFunctionalSkills } = req.session
+    const { prisonNumber } = req.params
+    this.addCurrentPageToHistory(req, `/prisoners/${prisonNumber}/induction/want-to-add-qualifications`)
```
Comment (nathanrussell-moj-digital) [inline server/routes/induction/common/wantToAddQualificationsController.ts:23]: I'm assuming this is not in an `if` block as per the other controllers because this page (Do they want to add qualifications?) is only accessible as part of changing the induction from long to short , and that you cannot access this question/page via a "Change" link on the main Education & Training or Work & Interests pages?

File: server/routes/induction/common/previousWorkExperienceDetailController.ts (hunk 8 above)
Comment (mattwills-moj-digital) [reply server/routes/induction/common/previousWorkExperienceDetailController.ts:37]: Indeed. As I'm sure you're aware, "e.g." literally means "for example" 😄 

The only difference would be the value within `updateInductionQuestionSet.hopingToWorkOnRelease`, but that's not relevant here.

File: server/routes/induction/common/reasonsNotToGetWorkController.ts [hunk 10]
```diff
@@ -18,6 +18,12 @@ export default abstract class ReasonsNotToGetWorkController extends InductionCon
   ): Promise<void> => {
     const { prisonerSummary, inductionDto } = req.session
 
+    // Check if we are in the midst of changing the main induction question set (i.e. from long route to short route)
```
Comment (mattwills-moj-digital) [inline server/routes/induction/common/reasonsNotToGetWorkController.ts:21]: Not "e.g." here as it's only relevant to short route. I might reword this to make it clearer.

**Review by nathanrussell-moj-digital — APPROVED** [review]
LGTM 👍

**Review by nathanrussell-moj-digital — APPROVED** [review]
👍
